### PR TITLE
Maximum number of Erlang processes and atoms can be configured via environment variables

### PR DESCRIPTION
**Pull request by gerhard**
## Proposed Changes

1. Allow maximum number of Erlang processes to be configured via an environment variable
1. Allow maximum number of Erlang atoms to be configured via an environment variable

## Types of Changes

- [ ] Bugfix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality) - partially relevant for #1513 
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

**Comment by gerhard** [comment]
Proposing to add this to `v3.6.x` & `v3.8.x`

**Comment by michaelklishin** [comment]
It can already be done using [RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS](http://www.rabbitmq.com/configure.html#define-environment-variables) but I have no objections to these two.

**Comment by lukebakken** [comment]
I'd say apply this to `v3.7.x` and `master` since `RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS` provides this function for `v3.6.x`.

**Comment by michaelklishin** [comment]
If this is driven by the needs of the BOSH release (which is a guess, perhaps @gerhard can elaborate), I have no objections in having this in 3.6.x as well.

**Comment by gerhard** [comment]
The BOSH release supports both `RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS` & specific properties, so it's not an issue using both. This PR is driven by making all properties in `SERVER_ERL_ARGS` equally configurable.

File: scripts/rabbitmq-env [hunk 1]
```diff
@@ -87,8 +87,14 @@ DEFAULT_SCHEDULER_BIND_TYPE="db"
 DEFAULT_DISTRIBUTION_BUFFER_SIZE=128000
 [ "x" = "x$RABBITMQ_DISTRIBUTION_BUFFER_SIZE" ] && RABBITMQ_DISTRIBUTION_BUFFER_SIZE=${DEFAULT_DISTRIBUTION_BUFFER_SIZE}
 
+DEFAULT_MAX_NUMBER_OF_PROCESSES=1048576
+[ "x" = "x$RABBITMQ_MAX_NUMBER_OF_PROCESSES" ] && RABBITMQ_MAX_NUMBER_OF_PROCESSES=="${DEFAULT_MAX_NUMBER_OF_PROCESSES}"
```
Comment (michaelklishin) [inline scripts/rabbitmq-env]: `==` seems to be bash-specific and is not available in zsh.

**Comment by lukebakken** [comment]
OK I tested my latest changes by creating a `rabbitmq-env.conf` file with these contents:

```
lbakken@shostakovich ~/issues/rabbitmq/rabbitmq-server/gh-1528
$ cat rabbitmq-env.conf 
SCHEDULER_BIND_TYPE=u
DISTRIBUTION_BUFFER_SIZE=65536
MAX_NUMBER_OF_PROCESSES=131072
MAX_NUMBER_OF_ATOMS=196608
```

Then running from this branch with this command:

```
make RABBITMQ_NODENAME='rabbit@shostakovich' RABBITMQ_CONF_ENV_FILE=$HOME/issues/rabbitmq/rabbitmq-server/gh-1528/rabbitmq-env.conf run-broker
```

`ps -ef` output shows that the settings were applied:

```
$ ps -ef|fgrep beam.smp
lbakken  25122 22593 15 11:55 pts/2    00:00:02 /home/lbakken/development/erlang/installs/19.3/erts-8.3/bin/beam.smp -W w -A 128 -P 131072 -t 196608 -stbt u -zdbbl 65536 -K true -- -root /home/lbakken/development/erlang/installs/19.3 -progname erl -- -home /home/lbakken -- -kernel shell_history enabled -pa /home/lbakken/development/rabbitmq/rabbitmq-server/ebin -s rabbit boot -sname rabbit@shostakovich -boot start_sasl -config /tmp/rabbitmq-test-instances/test -kernel inet_default_connect_options [{nodelay,true}] -sasl errlog_type error -sasl sasl_error_logger false -rabbit lager_log_root "/tmp/rabbitmq-test-instances/rabbit@shostakovich/log" -rabbit lager_default_file "/tmp/rabbitmq-test-instances/rabbit@shostakovich/log/rabbit@shostakovich.log" -rabbit lager_upgrade_file "/tmp/rabbitmq-test-instances/rabbit@shostakovich/log/rabbit@shostakovich_upgrade.log" -rabbit enabled_plugins_file "/tmp/rabbitmq-test-instances/rabbit@shostakovich/enabled_plugins" -rabbit plugins_dir "/home/lbakken/development/rabbitmq/rabbitmq-server/plugins" -rabbit plugins_expand_dir "/tmp/rabbitmq-test-instances/rabbit@shostakovich/plugins" -os_mon start_cpu_sup false -os_mon start_disksup false -os_mon start_memsup false -mnesia dir "/tmp/rabbitmq-test-instances/rabbit@shostakovich/mnesia/rabbit@shostakovich" -kernel inet_dist_listen_min 25672 -kernel inet_dist_listen_max 25672
```

Testing Windows next.

**Comment by lukebakken** [comment]
Tested on Windows 8.1 using [this alpha build](https://bintray.com/rabbitmq/all-dev/download_file?file_path=rabbitmq-server%2F3.7.4-alpha.12%2Frabbitmq-server-3.7.4-alpha.12.exe) with all of the `*.bat` files copied from this branch into `sbin`. After doing the copy, you must run `rabbitmq-service.bat remove` then `rabbitmq-service.bat install` for `erlsrv.exe` to be installed with the correct arguments for `erl.exe. Contents of `rabbitmq-conf-env.bat`:

```
set SCHEDULER_BIND_TYPE=u
set DISTRIBUTION_BUFFER_SIZE=65536
set MAX_NUMBER_OF_PROCESSES=131072
set MAX_NUMBER_OF_ATOMS=196608
```

File: scripts/rabbitmq-env (hunk 1 above)
Comment (gerhard) [reply scripts/rabbitmq-env]: My bad, definitely a typo.

**Comment by gerhard** [comment]
We used to have the `RABBITMQ_` prefix since #873. The sudo logic used to be:

```
if RABBITMQ_SCHEDULER_BIND_TYPE env is set
then use this value in SERVER_ERL_ARGS
otherwise set RABBITMQ_SCHEDULER_BIND_TYPE to DEFAULT_SCHEDULER_BIND_TYPE
```

Was this incorrect?

**Comment by michaelklishin** [comment]
@gerhard [the docs](http://www.rabbitmq.com/configure.html#define-environment-variables) suggest `rabbitmq-env.conf` should contain `var`, which takes precedence over `DEFAULT_var`. Both are overriden by `RABBITMQ_var` from the environment. This logic has been around for years. #873 therefore is incomplete.

**Comment by gerhard** [comment]
@michaelklishin got it. So we basically need to support both `RABBITMQ_var_name` & `var_name` defined in `rabbitmq-env.conf`. I will update the tests to cover both cases.

**Comment by gerhard** [comment]
@michaelklishin should be ready to merge.

**Pull request merged**